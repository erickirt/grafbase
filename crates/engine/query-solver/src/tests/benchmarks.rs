use crate::assert_solving_snapshots;

const SCHEMA: &str = r#"
enum join__Graph {
  SUB0 @join__graph(name: "sub0", url: "http://localhost:1000")
  SUB1 @join__graph(name: "sub1", url: "http://localhost:1001")
  SUB2 @join__graph(name: "sub2", url: "http://localhost:1002")
}

type Query @join__type(graph: SUB0) @join__type(graph: SUB1) @join__type(graph: SUB2) {
    node: Node
}
type Node @join__type(graph: SUB0, key: "id0") @join__type(graph: SUB1, key: "id0") @join__type(graph: SUB2, key: "id0") {
    id0: ID!
    n0: Node @join__field(graph: SUB1) @join__field(graph: SUB2)
    f0: String @join__field(graph: SUB1) @join__field(graph: SUB2)
    n1: Node @join__field(graph: SUB0) @join__field(graph: SUB2)
    f1: String @join__field(graph: SUB0) @join__field(graph: SUB2)
    n2: Node @join__field(graph: SUB0) @join__field(graph: SUB1)
    f2: String @join__field(graph: SUB0) @join__field(graph: SUB1)
}
"#;

#[tokio::test]
async fn alternating_subgraphs_1() {
    assert_solving_snapshots!(
        "alternating_subgraphs_1",
        SCHEMA,
        r#"
        query {
          node {
            n0 {
              n0 {
                f0
                f1
                f2
              }
              n2 {
                f0
                f1
                f2
              }
            }
          }
        }
        "#
    );
}
