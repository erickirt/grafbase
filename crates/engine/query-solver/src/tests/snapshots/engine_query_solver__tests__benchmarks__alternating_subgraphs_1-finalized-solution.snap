---
source: crates/engine/query-solver/src/tests/benchmarks.rs
expression: "digraph {\n    0 [ label = \"root\" ]\n    1 [ label = \"Root#sub1\", color=royalblue,shape=parallelogram ]\n    2 [ label = \"Query.node\" ]\n    3 [ label = \"Node.n0\" ]\n    4 [ label = \"Node.n0\" ]\n    5 [ label = \"Node.f0\" ]\n    6 [ label = \"Node.f2\" ]\n    7 [ label = \"FedEntity#sub2\", color=royalblue,shape=parallelogram ]\n    8 [ label = \"Node.n0\" ]\n    9 [ label = \"Node.f1\" ]\n    10 [ label = \"Node.n2\" ]\n    11 [ label = \"Node.f0\" ]\n    12 [ label = \"FedEntity#sub2\", color=royalblue,shape=parallelogram ]\n    13 [ label = \"Node.f1\" ]\n    14 [ label = \"Node.f2\" ]\n    15 [ label = \"*Node.id0\" ]\n    16 [ label = \"*Node.id0\" ]\n    0 -> 1 [ color=royalblue,fontcolor=royalblue ]\n    12 -> 15 [ color=orangered,arrowhead=inv ]\n    2 -> 3 [  ]\n    3 -> 4 [  ]\n    4 -> 5 [  ]\n    4 -> 6 [  ]\n    3 -> 7 [ color=royalblue,fontcolor=royalblue ]\n    1 -> 2 [  ]\n    8 -> 9 [  ]\n    3 -> 10 [  ]\n    10 -> 11 [  ]\n    10 -> 12 [ color=royalblue,fontcolor=royalblue ]\n    7 -> 8 [  ]\n    10 -> 14 [  ]\n    10 -> 15 [  ]\n    3 -> 16 [  ]\n    7 -> 16 [ color=orangered,arrowhead=inv ]\n    12 -> 13 [  ]\n}\n"
---
digraph {
    0 [ label = "root" ]
    1 [ label = "Root#sub1" ]
    2 [ label = "Query.node" ]
    3 [ label = "Node.n0" ]
    4 [ label = "Node.n0" ]
    5 [ label = "Node.f0" ]
    6 [ label = "Node.f2" ]
    7 [ label = "FedEntity#sub2" ]
    8 [ label = "Node.n0" ]
    9 [ label = "Node.f1" ]
    10 [ label = "Node.n2" ]
    11 [ label = "Node.f0" ]
    12 [ label = "FedEntity#sub2" ]
    13 [ label = "Node.f1" ]
    14 [ label = "Node.f2" ]
    15 [ label = "*Node.id0" ]
    16 [ label = "*Node.id0" ]
    0 -> 1 [ label = "QueryPartition" ]
    12 -> 15 [ label = "RequiredBySubgraph" ]
    2 -> 3 [ label = "Field" ]
    3 -> 4 [ label = "Field" ]
    4 -> 5 [ label = "Field" ]
    4 -> 6 [ label = "Field" ]
    3 -> 7 [ label = "QueryPartition" ]
    1 -> 2 [ label = "Field" ]
    8 -> 9 [ label = "Field" ]
    3 -> 10 [ label = "Field" ]
    10 -> 11 [ label = "Field" ]
    10 -> 12 [ label = "QueryPartition" ]
    7 -> 8 [ label = "Field" ]
    10 -> 14 [ label = "Field" ]
    10 -> 15 [ label = "Field" ]
    3 -> 16 [ label = "Field" ]
    7 -> 16 [ label = "RequiredBySubgraph" ]
    12 -> 13 [ label = "Field" ]
}
